### PR TITLE
Graphics/ctp 26 reimplement simple optimizations

### PR DESCRIPTION
**Pull request by benblack90**
I have added sphere-based frustum culling to the renderer. This involved adding a requirement to the RenderObject constructors for including a radius. 

As I'm still waiting for the object changes to come through before I can finish the other task of adding node sorting, I want to merge this branch now, so folks can merge from main, and get my changes.

### CODE REVIEW

**Review by Nijelous — APPROVED** [review]
Everything is A-ok, seems like it runs fine too, we may have to deal with the shadow stuff later.

File: CSC8503/GameTechRenderer.cpp [hunk 1]
```diff
@@ -146,7 +149,7 @@ void GameTechRenderer::BuildObjectList() {
 }
 
 void GameTechRenderer::SortObjectList() {
-
+	//std::sort(activeObjects.begin(), activeObjects.end(), )
```
Comment (erendgrmnc) [inline CSC8503/GameTechRenderer.cpp:152]: Best to remove commented out line before merging

File: CSC8503/GameTechRenderer.cpp [hunk 2]
```diff
@@ -136,8 +139,8 @@ void GameTechRenderer::BuildObjectList() {
 	gameWorld.OperateOnContents(
 		[&](GameObject* o) {
 			if (o->IsActive()) {
-				const RenderObject* g = o->GetRenderObject();
-				if (g) {
+			const RenderObject* g = o->GetRenderObject();
+				if (g && mFrameFrustum.SphereInsideFrustum(o->GetTransform().GetPosition(), o->GetRenderObject()->GetCullSphereRadius())) {
```
Comment (erendgrmnc) [inline CSC8503/GameTechRenderer.cpp:143]: I prefer declaring a variable and check them in if statements rather then complex logic controls.

And maybe we can change the names of the variable g and o, give them a much understandable names .

```suggestion
//Probably not the best variable name :)
bool isActivable = g && mFrameFrustum.SphereInsideFrustum(o->GetTransform().GetPosition(), o->GetRenderObject()->GetCullSphereRadius());
				if (isActivable ) {
```

**Review by erendgrmnc — APPROVED** [review]
I added some requests. If you think they are not correct or not necessary we can discuss them. Thanks for your work 🕺

File: CSC8503CoreClasses/RenderObject.h [hunk 3]
```diff
@@ -45,12 +45,17 @@ namespace NCL {
...
 		protected:
 			Mesh*		mesh;
 			Texture*	texture;
 			Shader*		shader;
 			Transform*	transform;
 			Vector4		colour;
+			float		mCullSphereRadius;
 		};
 	}
 }
```
Comment (Chenziyes) [inline CSC8503CoreClasses/RenderObject.h:61]: Every change looks like no problem.
But I don't know how to pull and run the code that needs to be reviewed.
@Eren, It's up to you.

File: CSC8503/TutorialGame.cpp [hunk 4]
```diff
@@ -287,7 +287,7 @@ GameObject* TutorialGame::AddFloorToWorld(const Vector3& position, const std::st
 		.SetScale(floorSize * 2)
 		.SetPosition(position);
 
-	floor->SetRenderObject(new RenderObject(&floor->GetTransform(), cubeMesh, basicTex, basicShader));
+	floor->SetRenderObject(new RenderObject(&floor->GetTransform(), cubeMesh, basicTex, basicShader, 120));
```
[the remark this answers is not in the document]
Comment (benblack90) [reply CSC8503/TutorialGame.cpp:290]: This is just to get it compiling: i think we're gonna be deleting all of this gameplay code in tutorial game anyway, tbh. But if we are keeping it, absolutely yes.

**Pull request merged**